### ISSUE TITLE
[@nativescript/facebook] iOS login error after redirecting back to Nativescript app

### ISSUE THREAD

**Issue by meganedesigns** [NativeScript/plugins#290]
I am getting the following error when using the nativescript/facebook (v2.0.0) package for facebook single sign on (currently just testing iOS).

```
*** Terminating app due to uncaught exception 'NativeScript encountered a fatal error: Uncaught TypeError: FBSDKApplicationDelegate.sharedInstance.applicationOpenURLOptions is not a function
at
FacebookAppDelegateImpl.applicationOpenURLOptions(file: src/webpack:/MyHighland/node_modules/@nativescript/facebook/index.ios.js:57:0)
', reason: '(null)'
```

When I initiate the `LoginManager.logInWithPermissions` function, I get the system dialog that "App Name" wants to use "facebook.com" to Sign In, and I then authorize in the facebook app. However when I am redirected back to my app, it crashes and I see the error above.

Do I need to manually create `AppDelegate.swift` / `SceneDelegate.swift` files as shown in step 3 of Facebook's iOS setup? And if I need to create them, is there any documentation or reference on how to do so in a nativescript application?https://developers.facebook.com/docs/ios/getting-started#step-3--connect-the-app-delegate 

```
"dependencies": {
    "@nativescript/background-http": "^5.0.2",
    "@nativescript/core": "~8.2.0",
    "@nativescript/facebook": "^2.0.0",
    "@nativescript/firebase-analytics": "2.0.1",
    "@nativescript/firebase-core": "2.0.1",
    "@nativescript/firebase-messaging": "2.0.1",
    "@nativescript/imagepicker": "^1.0.6",
    "@nativescript/local-notifications": "^6.0.0",
    "@nativescript/theme": "~3.0.1",
    "@triniwiz/nativescript-pager": "^13.0.3",
    "@triniwiz/nativescript-socketio": "^4.0.1",
    "@triniwiz/nativescript-toasty": "^4.1.3",
    "@vue/devtools": "^5.3.4",
    "axios": "^0.24.0",
    "highlight-words-core": "^1.2.2",
    "nativescript-appversion": "github:imandyie/nativescript-appversion",
    "nativescript-fonticon": "^2.0.2",
    "nativescript-gif": "^5.0.0",
    "nativescript-phone": "^3.0.2",
    "nativescript-ui-listview": "^10.0.2",
    "nativescript-vue": "~2.9.0",
    "nativescript-vue-devtools": "^1.5.1",
    "rxjs": "^7.5.5",
    "v-mask": "^2.3.0",
    "vue-axios": "^3.4.0",
    "vuex": "^3.6.2"
  },
  "devDependencies": {
    "@nativescript/android": "8.2.2",
    "@nativescript/ios": "~8.2.0",
    "@nativescript/webpack": "~5.0.0",
    "assert": "^2.0.0",
    "browserify-zlib": "^0.2.0",
    "buffer": "^6.0.3",
    "https-browserify": "^1.0.0",
    "nativescript-vue-template-compiler": "~2.9.0",
    "os-browserify": "^0.3.0",
    "path": "^0.12.7",
    "platform": "^1.3.6",
    "sass": "^1.32.8",
    "stream-browserify": "^3.0.0",
    "stream-http": "^3.2.0",
    "tns-ios": "^6.5.5",
    "tty-browserify": "0.0.1",
    "url": "^0.11.0",
    "util": "^0.12.4"
  }
```

**Comment by meganedesigns** [comment]
This only happens if the user has logged out of Facebook in Safari but has the Facebook app.

**Scenario 1:** User logged in to Facebook in Safari, so they are not redirected to the Facebook app. Facebook app is installed on the device.

https://user-images.githubusercontent.com/8839452/175794850-eab5f8a0-bb86-4a11-b77c-f5ca1cc1f7a1.MOV


**Scenario 2:** User has the Facebook app installed, but has logged out of Facebook in Safari (or the Safari login has expired). The user is sent to the Facebook app to authorize, but after being redirected back to the Nativescript app, the app crashes and I see the console error below.

https://user-images.githubusercontent.com/8839452/175967687-003543a8-f61e-44fe-969b-6551ea859957.mov



 In the console, I see this error:
```
*** Terminating app due to uncaught exception 'NativeScript encountered a fatal error: Uncaught TypeError: FBSDKApplicationDelegate.sharedInstance.applicationOpenURLOptions is not a function
at
FacebookAppDelegateImpl.applicationOpenURLOptions(file: src/webpack:/MyHighland/node_modules/@nativescript/facebook/index.ios.js:57:0)
', reason: '(null)'
```

**Comment by NathanWalker** [comment]
@meganedesigns in your app, do you have a custom delegate wired up already? just curious. I also noticed facebook released 14.0.0 of the accompanying sdk's (current version is using 13.1.0); we may try updating those on a alpha for you to try as well.

**Comment by NathanWalker** [comment]
Additionally are you calling `LoginManager.init()` in your main.ts or elsewhere?

**Comment by meganedesigns** [comment]
@NathanWalker No custom delegate wired up anywhere. And yes, I'm calling `LoginManager.init()` in our main.js file.

**Comment by NathanWalker** [comment]
@meganedesigns when possible if you could do the following:
```
code node_modules/@nativescript/facebook/index.ios.js
```
To open that file in VS Code. Then comment out these lines:
```
//FacebookAppDelegateImpl.prototype.applicationOpenURLOptions = function (app, url, options) {
//        return FBSDKApplicationDelegate.sharedInstance.applicationOpenURLOptions(app, url, options);
//};
```
Save and then run the app again. I'm not seeing that here so curious what could affect it from your side but let's try a few things like that and it may reveal an answer. That will either get farther to reveal something else of note or simply return to app and do nothing (but with no crash).

**Comment by meganedesigns** [comment]
I had to comment out the function afterward to prevent the app crash.

```
    // FacebookAppDelegateImpl.prototype.applicationOpenURLOptions = function (app, url, options) {
    //     return FBSDKApplicationDelegate.sharedInstance.applicationOpenURLOptions(app, url, options);
    // };
    // FacebookAppDelegateImpl.prototype.applicationOpenURLSourceApplicationAnnotation = function (application, url, sourceApplication, annotation) {
    //     return FBSDKApplicationDelegate.sharedInstance.applicationOpenURLSourceApplicationAnnotation(application, url, sourceApplication, annotation);
    // };
```

Error was similar to the previous one.
```
*** Terminating app due to uncaught exception 'NativeScript encountered a fatal error: Uncaught TypeError: FBSDKApplicationDelegate.sharedInstance.applicationOpenURLSourceApplicationAnnotation is not a function
at
FacebookAppDelegateImpl.applicationOpenURLSourceApplicationAnnotation(file: src/webpack:/MyHighland/node_modules/@nativescript/facebook/index.ios.js:60:0)
', reason: '(null)'
```

With both functions above commented out, I am redirected back to my app with the in-app browser still active, and no response/error.
```
async facebookLogin() {
    await LoginManager.logInWithPermissions(['public_profile']).then(profile => {
        console.log('profile', profile);
    }).catch(e => {
        console.log('Facebook login error:', e);
    });
}
```

**Comment by meganedesigns** [comment]
@NathanWalker is there a way for me to try out the update in the PR? I'm not seeing an alpha release for me to try.

**Comment by NathanWalker** [comment]
Hi @meganedesigns yes absolutely, please try `2.1.0-alpha.0` - can set your package directly to that or use the `alpha` tag.
Be sure to `ns clean`!

Lemme know - I'm seeing good behavior here now.

**Comment by meganedesigns** [comment]
@NathanWalker The alpha version didn't help in my project, but I created a fresh build with the bare-bones packages need to run a NS app. That works, so I'm going to add/remove packages until I find the issue.

**Comment by NathanWalker** [comment]
Curious indeed @meganedesigns  - if you'd like to invite me to take a peek on the project I'd be curious to have a look as well. I tested with the demo app setup in this workspace using our FB account on both conditions with good results here but it's possible some other plugin in use may be affecting, lemme know.

**Comment by NathanWalker** [comment]
My guess is there is another delegate in use would have to explain it. If you see here:
https://github.com/NativeScript/plugins/blob/f314b174dd680ec76fa8b8bb42d50003b69c98ab/packages/facebook/index.ios.ts#L232-L234
The delegate is only attached if there isn't a previous delegate configured which is a bit naive tbh. Your app via another plugin is likely attaching a delegate (firebase usually) and if we prove that's the case we can probably improve this bit more to simply expose the FB delegate that you could wire in or allow it to mix in.

**Comment by meganedesigns** [comment]
@NathanWalker I was just coming on here to say that I figured out the conflict was with Firebase.

**Comment by NathanWalker** [comment]
We can probably keep this open to ensure it doesn’t cause issue with different firebase setups  - any other detail could share with how your main is setup and/or how firebase is integrated and we could make further adjustments to consider the case.

**Comment by meganedesigns** [comment]
I am using the following Firebase packages:

```
"@nativescript/firebase-analytics": "2.0.1",
"@nativescript/firebase-core": "2.0.1",
"@nativescript/firebase-messaging": "2.0.1",
```

I will try to do a little more digging today to see which (possibly all) package(s) add App Delegates.

**Comment by gelzis** [comment]
Have the same problem (if user is using mobile app for the auth it always fails, works fine with the auth coming from the web), I made changes for this plugin to preserve everything that was used before, took approach thats used here: https://github.com/nativescript-community/universal-links/blob/master/src/universal-links/index.ios.ts#L20 that it extends previous handlers, but this doesn't help. Basically `FBSDKApplicationDelegate.sharedInstance.applicationOpenURLOptions` never exists in the context of the `applicationOpenURLOptions` handler call. There is no timing issue also, putting in an interval for checking it existence shows that it never exists also afterwards.

**Comment by meganedesigns** [comment]
@NathanWalker firebase-core and firebase-messaging both add to the AppDelegate. Interestingly when I try to do the Firebase/Facebook auth outlined here on a fresh app install, I get the same problem as I do with my app: https://docs.nativescript.org/plugins/firebase-auth.html#other-sign-in-methods

**Comment by meganedesigns** [comment]
@NathanWalker sorry for all the tags, but I setup a test repo with Firebase and Facebook login for you to try out. I have removed our Facebook/Firebase keys, but I can setup some test accounts if you need it.

https://github.com/meganedesigns/myCoolApp

**Comment by NathanWalker** [comment]
Thank you so much for taking the time to setup that sample @meganedesigns it's been super helpful - still looking into it but I plan on having a solution at least by early next week; just wanted make sure I consider the cleanest way forward.

**Comment by NathanWalker** [comment]
@meganedesigns resolved for all cases in `2.1.0`, lemme know if any further issues. Thank you again for the great sample here.
